bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/handler/ZonePlayerHandler.java

```diff
@@ -1353,11 +1312,118 @@ && hasValueChanged(albumArtURI, memberHandler.stateMap.get("CurrentAlbumArtURI")
+                }
+            }
+        }
+        if (!opmlUrlSucceeded) {
+            if (radioTitle != null && !radioTitle.isEmpty()) {
+                title = radioTitle;
+                information = title;
+                if (trackStreamContent != null && !trackStreamContent.isEmpty()) {
+                    information += " - " + trackStreamContent;
+                }
+                needsUpdate = true;
+            }
+        }
+        return new ResultMediaParsing(null, null, title, information, stationId, needsUpdate);
+    }
+
+    private ResultMediaParsing parseRadioAppMediaInfo(@Nullable String radioTitle,
+            @Nullable String trackStreamContent) {
+        String artist = null;
+        String album = null;
+        String title = null;
+        String information = null;
+        boolean needsUpdate = false;
+        if (radioTitle != null && !radioTitle.isEmpty()) {
+            title = radioTitle;
+            information = title;
+            if (trackStreamContent != null && !trackStreamContent.isEmpty()) {
+                String[] contents = trackStreamContent.split("\\|");
+                String contentTitle = null;
+                for (int i = 0; i < contents.length; i++) {
+                    if (contents[i].startsWith("TITLE ")) {
+                        contentTitle = contents[i].substring(6).trim();
+                    }
+                    if (contents[i].startsWith("ARTIST ")) {
+                        artist = contents[i].substring(7).trim();
+                    }
+                    if (contents[i].startsWith("ALBUM ")) {
+                        album = contents[i].substring(6).trim();
+                    }
+                }
+                if ((artist == null || artist.isEmpty()) && contentTitle != null && !contentTitle.isEmpty()
+                        && !contentTitle.startsWith("Advertisement_")) {
+                    // Try to extract artist and song title from contentTitle
+                    int idx = contentTitle.indexOf(" - ");
+                    if (idx > 0) {
+                        artist = contentTitle.substring(0, idx);
+                        title = contentTitle.substring(idx + 3);
+                    }
+                }
+                if (artist != null && !artist.isEmpty()) {
+                    information += " - " + artist;
+                }
+                if (album != null && !album.isEmpty()) {
+                    information += " - " + album;
+                }
+                if (!radioTitle.equals(title)) {
+                    information += " - " + title;
+                } else if (contentTitle != null && !contentTitle.isEmpty()
+                        && !contentTitle.startsWith("Advertisement_")) {
+                    information += " - " + contentTitle;
+                }
+            }
+            needsUpdate = true;
+        }
+        return new ResultMediaParsing(artist, album, title, information, null, needsUpdate);
+    }
```

Comment:
If you want to go further, this could be moved into class `ResultMediaParsing` as a public static method returning the result object. The class could then be extracted as a separate public class, which would then be easy to implement unit tests for. I understand that this change was triggered by a small proposal from me, so just treat this comment as an additional idea which may grow on you or not. ;-)

Reply:
This is done !